### PR TITLE
#447 Mongo Exceptions Normalization

### PR DESCRIPTION
**Pull request by muhammadhamzasajjad**
This normalizes mongo exceptions as in #584. The logic for exception normalization is now moved to  `mongo_storage.cpp`  as we also support a `MockMongoClient` which can mock failures. The Permission exceptions are also normalized. Previously there were two types of permission exceptions. One was triggered when a storage API returned a permission failure, the second was triggered if an unpermitted storage operation was attempted as per the library `OpenMode`. The two permission have now been normalized to inherit from the same class.

Furthermore, the exception hierarchy in python has also been changed. `DuplicateKeyException` inherits from `StorageException` in C++, this should be reflected on python side too. Why? because if someone wanted they could catch all the storage exceptions as follows instead of having to catch `PermissionException`, `DuplicateKeyException` and `StorageException` separately:

```python
try:
  # an operation that calls storage from python
except StorageException:
  # log storage error
```

#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [ ] Have you updated the relevant docstrings, documentation and copyright notice?
 - [ ] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [ ] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [ ] Are API changes highlighted in the PR description?
 - [ ] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

### CODE REVIEW

File: cpp/arcticdb/python/python_module.cpp [hunk 1]
```diff
@@ -236,6 +236,9 @@ void register_error_code_ecosystem(py::module& m, py::exception<arcticdb::Arctic
             internal_exception(msg.c_str());
         }
     });
+    // DuplicateKeyException and PermissionException are subclasses of StorageException so keep the hierarchy in Python too
```
Comment (poodlewars) [inline cpp/arcticdb/python/python_module.cpp:239]: drop the comment?

File: cpp/arcticdb/storage/python_bindings.cpp [hunk 2]
```diff
@@ -34,7 +34,7 @@ std::shared_ptr<LibraryIndex> create_library_index(const std::string &environmen
     return std::make_shared<LibraryIndex>(EnvironmentName{environment_name}, mem_resolver);
 }
 
-void register_bindings(py::module& storage, py::exception<arcticdb::ArcticException>& base_exception) {
+void register_bindings(py::module& storage, py::exception<arcticdb::ArcticException>&) {
```
Comment (poodlewars) [inline cpp/arcticdb/storage/python_bindings.cpp:37]: Remove the unused argument completely?

File: cpp/arcticdb/storage/test/test_storage_exceptions.cpp
Comment (poodlewars) [inline cpp/arcticdb/storage/test/test_storage_exceptions.cpp:1]: It's really great working like this where you've written the tests with the old behaviour in an earlier PR. Makes it much easier to see that the changes are tested, and what behaviour changes you're introducing :+1:

**Review by poodlewars — CHANGES_REQUESTED** [review]
This looks great! Just minor comments

File: cpp/arcticdb/python/python_module.cpp (hunk 1 above)
Comment (muhammadhamzasajjad) [reply cpp/arcticdb/python/python_module.cpp]: Done in [59436ce](https://github.com/man-group/ArcticDB/pull/1411/commits/59436ceca3bb044cf4e6412c2eceaebf9080880f)

File: cpp/arcticdb/storage/python_bindings.cpp (hunk 2 above)
Comment (muhammadhamzasajjad) [reply cpp/arcticdb/storage/python_bindings.cpp]: Done in [59436ce](https://github.com/man-group/ArcticDB/pull/1411/commits/59436ceca3bb044cf4e6412c2eceaebf9080880f)

**Review by poodlewars — APPROVED** [review]
Looks good! Please squash before you merge this in

**Pull request merged**